### PR TITLE
GPU support for complex operators

### PR DESCRIPTION
**Pull request by stefanhenneking**
Features:
- [x] Enable support for device for SesquilinearForm with PA
- [x] Enable support for device for ComplexGridFunction
- [x] Enable support for device for ComplexLinearForm
- [x] Add device option to the serial example ex22
- [x] Enable support for device for ParSesquilinearForm with PA
- [x] Enable support for device for ParComplexGridFunction
- [x] Enable support for device for ParComplexLinearForm
- [x] Add device option to the parallel example ex22p
- [x] Update `CHANGELOG`

Remarks:
- This PR depends on: #1563 and #1468
- One issue that complicates this feature is described in #1578

Timings (obtained on [Ray](https://hpc.llnl.gov/hardware/platforms/Ray)):
1. `ex22 -m ../data/inline-hex.mesh -no-vis -r 2 -o 1 -p 2`

| Timing            | `-pa`         | `-pa -d cuda` |
| -------------     | ------------- | ------------- |
| One Operator Mult | `0.025 sec`    | `0.00017 sec`    |
| GMRES (no prec.)  | `1.80 sec`    | `0.13 sec`    |
| Entire run (no prec.) | `2.25 sec`    | `0.76 sec`    |

2. `lrun -n 4 ex22p -m ../data/inline-hex.mesh -no-vis -o 1 -p 2`

| Timing            | `-pa`         | `-pa -d cuda` |
| -------------     | ------------- | ------------- |
| FGMRES (no prec.)  | `0.90 sec`    | `0.14 sec`    |
| Entire run (no prec.) | `1.03 sec`    | `0.49 sec`    |

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1575](https://github.com/mfem/mfem/pull/1575) | @stefanhenneking | @tzanio | @mlstowell + @psocratis + @dylan-copeland | 07/27/20 | 08/18/20 | 08/25/20 | |

### CODE REVIEW

**Comment by stefanhenneking** [comment]
@mlstowell suggested in https://github.com/mfem/mfem/pull/1598#pullrequestreview-459401337 that it would be convenient to have (public) methods to easily sync the real and imaginary parts of `[Par]ComplexGridFunction` and `[Par]ComplexLinearForm`. I added methods for that in `complex_fem.hpp`. However, in `complex_fem.cpp`, I personally prefer to explicitly call the `SyncMemory`/`SyncAliasMemory` methods on the member variables instead of using the public method (I think it's easier to read this way), so I left those calls untouched. These new methods are named `Sync()` and `SyncAlias()`, similar to the corresponding functions on memory objects, but I am open to suggestions for different names.

File: fem/plinearform.hpp [hunk 1]
```diff
@@ -92,6 +92,25 @@ class ParLinearForm : public LinearForm
...
 
+
+   /** @brief Make the ParLinearForm reference external data on a new
+       FiniteElementSpace. */
+   /** This method changes the FiniteElementSpace associated with the ParLinearForm
+       to @a *f and sets the data of the Vector @a v (plus the @a v_offset) as external
+       data in the ParLinearForm.
+       @note This version of the method will also perform bounds checks when
+       the build option MFEM_DEBUG is enabled. */
+   virtual void MakeRef(FiniteElementSpace *f, Vector &v, int v_offset);
```
Comment (psocratis) [inline fem/plinearform.hpp:103]: Hey @stefanhenneking, 

what is the purpose of this method (with this particular signature)? Is it really needed ?

Comment (psocratis) [reply fem/plinearform.hpp:103]: @stefanhenneking , never mind...disregard the previous question. 
However, shouldn't there be a check that the dynamic cast is not null, i.e, pfes is a ParFiniteElementSpace?

Comment (stefanhenneking) [reply fem/plinearform.hpp:103]: Good catch! I'll add an assert statement to check that `pfes` is not `NULL` after casting.

File: examples/ex22p.cpp [hunk 2]
```diff
@@ -13,10 +13,10 @@
 //               mpirun -np 4 ex22p -m ../data/inline-hex.mesh -o 2 -p 2
 //               mpirun -np 4 ex22p -m ../data/star.mesh -o 2 -sigma 10.0
 //
-// With partial assembly:
-//               mpirun -np 4 ex22p -m ../data/inline-quad.mesh -o 1 -p 1 -pa
-//               mpirun -np 4 ex22p -m ../data/inline-hex.mesh -o 1 -p 2 -pa
-//               mpirun -np 4 ex22p -m ../data/star.mesh -o 2 -sigma 10.0 -pa
+// Device sample runs:
+//               mpirun -np 4 ex22p -m ../data/inline-quad.mesh -o 1 -p 1 -pa -d cuda
```
Comment (psocratis) [inline examples/ex22p.cpp:17]: Should we leave maybe one sample run with -pa on the cpu? Same for ex22.

Comment (stefanhenneking) [reply examples/ex22p.cpp:17]: Sure, I can add a sample run with `-pa` to the "Sample Runs" on CPU.

**Comment by tzanio** [comment]
@mlstowell can you please review this?

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: examples/ex22p.cpp [hunk 3]
```diff
@@ -160,19 +164,24 @@ int main(int argc, char *argv[])
    ComplexOperator::Convention conv =
       herm_conv ? ComplexOperator::HERMITIAN : ComplexOperator::BLOCK_SYMMETRIC;
 
-   // 3. Read the (serial) mesh from the given mesh file on all processors. We
+   // 3. Enable hardware devices such as GPUs, and programming models such as
+   //    CUDA, OCCA, RAJA and OpenMP based on command line options.
+   Device device(device_config);
+   device.Print();
```
Comment (v-dobrev) [inline examples/ex22p.cpp:170]: I just noticed this in the internal tests -- this print should be called only on rank 0.

Comment (stefanhenneking) [reply examples/ex22p.cpp]: Thanks, @v-dobrev, for catching this. Fixed it now.

**Comment by stefanhenneking** [comment]
Also merged recent master to resolve conflict in `CHANGELOG`.

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Pull request merged**